servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -77,91 +80,120 @@ public void onComplete() {
+            tSubscription.delayedSubscription(new Subscription() {
+                @Override
+                public void request(final long n) {
+                    assert n == bufferSizeHint;
+                    itemsPending = (int) n;
+                    cs.request(n);
+                }
+
+                @Override
+                public void cancel() {
+                    cs.cancel();
+                }
+            });
         }
 
         @Override
         public void onNext(@Nullable final T t) {
-            assert itemsPending > 0;
-            --itemsPending;
-            state.accumulate(t);
-            if (itemsPending == 0) {
-                assert subscription != null;
-                itemsPending = bufferSizeHint;
-                subscription.request(bufferSizeHint);
+            final int cItemsPending = --itemsPending;
+            assert cItemsPending >= 0;
+            try {
+                state.accumulate(t, target);
+            } finally {
+                if (cItemsPending == 0 && state.requestMore()) {
+                    tSubscription.request(bufferSizeHint);
+                }
             }
         }
 
         @Override
         public void onError(final Throwable t) {
-            state.itemsTerminated(error(t));
+            state.itemsTerminated(error(t), target, bSubscription);
         }
 
         @Override
         public void onComplete() {
-            state.itemsTerminated(complete());
+            state.itemsTerminated(complete(), target, bSubscription);
         }
     }
 
     private static final class BoundariesSubscriber<T, B> implements Subscriber<Accumulator<T, B>> {
         private final State state;
         private final Subscriber<? super B> target;
+        private final DelayedSubscription bSubscription;
         private final Subscription tSubscription;
 
-        @Nullable
-        private ConcurrentSubscription subscription;
-
-        BoundariesSubscriber(final State state, final Subscriber<? super B> target, final Subscription tSubscription) {
+        BoundariesSubscriber(final State state, final Subscriber<? super B> target,
+                             final DelayedSubscription bSubscription, final Subscription tSubscription) {
             this.state = state;
             this.target = target;
+            this.bSubscription = bSubscription;
             this.tSubscription = tSubscription;
         }
 
         @Override
-        public void onSubscribe(final Subscription bSubscription) {
-            subscription = wrap(new Subscription() {
+        public void onSubscribe(final Subscription subscription) {
+            // We may cancel from multiple threads and DelayedSubscription will atomically swap if a cancel occurs but
+            // it will not prevent concurrent access between request(n) and cancel() on the original subscription.
+            bSubscription.delayedSubscription(ConcurrentSubscription.wrap(new Subscription() {
```

Comment:
`tSubscription` is already wrapped in `ConcurrentSubscription` did you consider just wrapping `subscription` here?